internal/metrics/daemon.go

```diff
@@ -0,0 +1,202 @@
+package metrics
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	workloadMetric = "workloadName"
+	scrapeTimeout  = 15 * time.Second
+)
+
+type ScrappeCallback func(data []model.Vector, labels map[string]string)
+
+type TargetMetric struct {
+	workloadName string
+	interval     time.Duration
+	urls         []string
+	scrappers    []*HTTPScraper
+	ctx          context.Context
+	cancel       context.CancelFunc
+	forcetrigger chan bool
+	lock         sync.RWMutex
+}
+
+func NewTargetMetric(workloadName string, interval time.Duration, urls []string) *TargetMetric {
+	return &TargetMetric{
+		workloadName: workloadName,
+		interval:     interval,
+		urls:         urls,
+		forcetrigger: make(chan bool),
+	}
+}
+func (tg *TargetMetric) ForceEvent() {
+	tg.forcetrigger <- true
+}
+
+func (tg *TargetMetric) Start(cb ScrappeCallback) {
+	log.Infof("started targetMetric for workload '%v' with urls '%+v'",
+		tg.workloadName, tg.urls)
+
+	tg.lock.Lock()
+	// if it was started early, stop it
+	if tg.cancel != nil {
+		tg.cancel()
+	}
+
+	tg.createScrapper()
+	tg.ctx, tg.cancel = context.WithCancel(context.Background())
+	ticker := time.NewTicker(tg.interval)
+	tg.lock.Unlock()
+
+	getdata := func() {
+		log.Debugf("ticker run for workload '%v'", tg.workloadName)
+		ctx, _ := context.WithTimeout(context.Background(), scrapeTimeout)
+		data := tg.Run(ctx)
+		cb(data, map[string]string{workloadMetric: tg.workloadName})
+	}
+	for {
+		select {
+		case <-tg.forcetrigger:
+			getdata()
+		case <-ticker.C:
+			getdata()
+		case <-tg.ctx.Done():
+			log.Debugf("time ticker for workload '%v' stopped", tg.workloadName)
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (tg *TargetMetric) IsStopped() bool {
+	return tg.cancel == nil
+}
+
+func (tg *TargetMetric) Stop() {
+	tg.lock.Lock()
+	defer tg.lock.Unlock()
+	if tg.cancel != nil {
+		tg.cancel()
+	}
+	tg.cancel = nil
+	return
+}
+
+func (tg *TargetMetric) createScrapper() {
+	scrappers := []*HTTPScraper{}
+	for _, url := range tg.urls {
+		scrappe, err := NewHTTPScraper(url)
+		if err != nil {
+			log.Errorf("cannot start scrapper for %v: %v", url, err)
+		}
+		scrappers = append(scrappers, scrappe)
+	}
+	tg.scrappers = scrappers
+}
+
+func (tg *TargetMetric) Run(ctx context.Context) []model.Vector {
+	res := []model.Vector{}
```

Comment:
WDYT about just appending to a `model.Vector` instead of having slice of de-facto slices?